### PR TITLE
Threading pool

### PR DESCRIPTION
**Pull request by jkbonfield**
See samtools/htslib#397 for discussions.

### CODE REVIEW

File: bam_sort.c [hunk 1]
```diff
@@ -1815,6 +1848,7 @@ int bam_sort_core_ext(int is_by_qname, const char *fn, const char *prefix,
         }
         if (bam_merge_core2(is_by_qname, fnout, modeout, NULL, n_files, fns,
                             MERGE_COMBINE_RG|MERGE_COMBINE_PG|MERGE_FIRST_CO,
+                            //MERGE_FIRST_CO | MERGE_NO_TRANS,
```
Comment (mp15) [inline bam_sort.c]: Probably want to remove this line as it GNDN

File: bam_sort.c [hunk 2]
```diff
@@ -1166,13 +1167,35 @@ int bam_merge_core2(int by_qname, const char *out, const char *mode,
...
+    }
+
     translation_tbl = (trans_tbl_t*)calloc(n, sizeof(trans_tbl_t));
     if (!translation_tbl) goto mem_fail;
     RG = (char**)calloc(n, sizeof(char*));
     if (!RG) goto mem_fail;
 
     // prepare RG tag from file names
-    if (flag & MERGE_RG) {
+    if ((flag & MERGE_RG)) {
```
Comment (mp15) [inline bam_sort.c]: Was this change to prevent a warning? It doesn't look necessary given you didn't do the same at 1171?

File: .travis.yml [hunk 3]
```diff
@@ -15,6 +15,18 @@ matrix:
     - os: osx
       compiler: clang
 
+# For linux systems
+addons:
```
Comment (mp15) [inline .travis.yml]: I think this bit should be promoted to it's own patch and landed asap.

File: bam_sort.c (hunk 2 above)
Comment (jkbonfield) [inline bam_sort.c]: No idea to be honest; it doesn't ring any bells, but it's been a while.   It could be removed and then see if warnings do indeed appear.

File: bam_sort.c (hunk 1 above)
Comment (jkbonfield) [inline bam_sort.c]: If we remove that we also want to remove the whole MERGE_NO_TRANS code elsewhere.  I recall working on this in merge so that it was faster and didn't spend a lot of pointless time translating IDs between files when infact all the IDs are identical because we just took the same header and spat it out N times.  This is simply and optimisation to remove that necessity.

Alas what I can't remember is why I didn't enable it.  I'm no expert in merge, so maybe it was just fear of the unknown or maybe it was a work in progress and something was amiss. Hmm.  Can you see anything wrong with the logic of what it's attempting to do?

Is this simply an issue of convolution between samtools merge and samtools sort?

**Comment by jkbonfield** [comment]
This is ancient and has cruft which I'm wary of pulling in immediately before a release.

I'll look through it and see what's necessary.  It may even be preferable to just create a new PR given the state this is in.

Technically it may not even be necessary in many cases now we have --input-fmt-option added to pretty much every tool, permitting e.g. ```--input-fmt-option nthreads=8```, but -@ is simpler.

**Comment by jkbonfield** [comment]
Closing in favour of PR #635
